enhancements/ingress/optional-ingress-hypershift.md

```diff
@@ -0,0 +1,320 @@
+
+#### Component routes
+
+[The component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) is an optional list of routes that are managed by OpenShift components that a cluster-admin is able to configure the hostname and serving certificate for.
+The two main OpenShift components which use the component routes are:
+[the web console](https://docs.openshift.com/container-platform/4.13/web_console/customizing-the-web-console.html#customizing-the-console-route_customizing-web-console) and [the authentication](https://docs.openshift.com/container-platform/4.13/authentication/configuring-internal-oauth.html#customizing-the-oauth-server-url_configuring-internal-oauth).
+The console and the authentication operators implement the component routes by watching [the Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html) and updating the corresponding routes.    
+The ingress operator creates the required role and role binding for the OpenShift component's service account so that it can get the custom service certificate secret from the `openshift-config` namespace.  
+In the absence of the ingress operator, the component routes:
+- cannot be customized by the owner components due to the lack of the RBAC
+- become not implemented due to [the absence of the default ingress controller](#drawbacks)
+
+The cluster instance administrator will still be able use [the Ingress Config API](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html)
+but the component owner (operator) won't be able to access the secret or rely on the route implementation.     
+This may be misleading for the cluster instance administrator. However the component routes were never fully implemented on HyperShift.
+For instance, the authentication operator which is supposed to implement the authentication component route is not used at all.
+
+#### Ingress to route controller
+
+[The Kubernetes Ingress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#ingress-v1-networking-k8s-io) on OpenShift is implemented through translation into routes.
+[The ingress-to-route controller](https://github.com/openshift/route-controller-manager#ingress-to-route-controller) is the component responsible for this functionality. When the default ingress controller is not present, the routes are left without serving, resulting in the ingresses also being unserved. On HyperShift this can be mitigated by [any ingress controller implementation](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/).
+In the context of this enhancement, [AWS Load Balancer Operator](/enhancements/ingress/aws-load-balancer-operator.md) is the expected implementation.
```

Comment:
Would it be reasonable to turn off the ingress-to-route controller when the capability is not enabled?  Perhaps as follow-up work?

Reply:
I guess "Add Route API to the ingress capability" does address my question, albeit obliquely:

> [...] In the situation when the ingress is disabled the choice lies in between the following options:
> 1. Keep the Route API and let the OpenShift components create routes which are not served by any ingress controller.
> 2. Keep the Route API and let the OpenShift components use an alternative ingress which supports the Route API.
> 3. Remove the Route API and leave the OpenShift components to adapt to the situation, which implies code changes to stop creating routes and to migrate to a different ingress API (e.g. Kubernetes Ingress).

I figured that "OpenShift components" referred to operators that created routes for their operands, but I understand now you also mean to include the ingress-to-route controller.  

This list sets up a false dichotomy; one unstated option would be to continue to let operators create their own routes but still turn off the ingress-to-route controller.  

> The first option seems to be the easiest (for the implementation), [...]

I suppose this (ease of implementation) is ultimately the reason why we don't turn off the ingress-to-route controller when the ingress capability is not enabled, right?  

> since the first option doesn't preclude the second one, it seems to be pragmatic to pick the first option, leaving open the possibility of implementing the second option later on.

I think it's a reasonable compromise to continue having the various operators (auth, console, monitoring) creating routes.  Those routes could even be useful as references when creating equivalent ingresses, should the cluster-admin choose to do so.  

However, leaving the ingress-to-route controller enabled seems less useful and more likely to mislead the typical project-admin who creates an ingress, observes that OpenShift generates a route for that ingress, and could be thereby misled to expect that OpenShift accept traffic for that ingress.  

I suppose there is the possibility that someone could turn off the ingress capability and then install a custom router deployment or even a third-party controller that implements the Route API, in which case the ingress-to-route controller _could_ still be useful.  That seems unlikely though.  

Anyway, I still think it would make sense to add a short statement here to explain why we keep translating ingresses to routes even on a cluster that doesn't serve routes, with a link to the "Add Route API to the ingress capability" section for the lengthier explanation.

Reply:
> I figured that "OpenShift components" referred to operators that created routes for their operands, but I understand now you also mean to include the ingress-to-route controller.

Yes. I believe I meant the ingres-to-route controller too.

> I suppose this (ease of implementation) is ultimately the reason why we don't turn off the ingress-to-route controller when the ingress capability is not enabled, right?

Right.

> However, leaving the ingress-to-route controller enabled seems less useful and more likely to mislead the typical project-admin who creates an ingress, observes that OpenShift generates a route for that ingress, and could be thereby misled to expect that OpenShift accept traffic for that ingress.

Yes, the translation of the ingresses to routes seems unnecessary in case the ingress capability is disabled.

Reply:
In order to highlight the discussed I added this phrase to "Ingress to route controller" risk chapter:

> It's also important to note that the routes generated from the ingress objects may become unnecessary if the ingress capability is disabled. However, this issue is not critical and can be considered for implementation as a follow-up item.